doc/api/module.md

```diff
@@ -1048,6 +1048,86 @@ returned object contains the following keys:
 * columnNumber: {number} The 1-indexed columnNumber of the
   corresponding call site in the original source
 
+## TypeScript support
+
+<!-- YAML
+added: REPLACEME
+-->
+
+> Stability: 1.0 - Early development
+
+The flag [`--experimental-strip-types`][] enables Node.js to run TypeScript
+files that contain only type annotations. Such files contain no TypeScript
+features that require transformation, such as enums or namespaces. Node.js will
+replace inline type annotations with whitespace, and no type checking is
+performed. TypeScript features that depend on settings within `tsconfig.json`,
+such as paths or converting newer JavaScript syntax to older standards, are
+intentionally unsupported.
+
+To get fuller TypeScript support, including support for enums and namespaces
+and paths, see <https://nodejs.org/en/learn/getting-started/nodejs-with-typescript>.
+The built-in TypeScript support is designed to be lightweight and as fast as
+possible. By intentionally not supporting syntaxes that require JavaScript code
+generation, and by replacing stripped types with whitespace, Node.js can run
+TypeScript code without the need for source maps and with less overhead.
+
+### Determining module system
+
+Node.js supports both [CommonJS][] and [ES Modules][] syntax in TypeScript
+files. Node.js will not convert from one module system to another; if you want
+your code to run as an ES module, you must use `import` and `export` syntax,
+and if you want your code to run as CommonJS you must use `require` and
+`module.exports`.
+
+* `.ts` files will have their module system determined
+  [the same way as `.js` files.][] To use `import` and `export` syntax, add
+  `"type": "module"` to the nearest parent `package.json`.
+* `.mts` files will always be run as ES modules, similar to `.mjs` files.
+* `.cts` files will always be run as CommonJS modules, similar to `.cjs` files.
+* `.tsx` files are unsupported.
+
+As in JavaScript files, [file extensions are mandatory][] in `import` statements
+and `import()` expressions: `import './file.ts'`, not `import './file'`.
+Because of backward compatibility, file extensions are also mandatory in
+`require()` calls: `require('./file.ts')`, not `require('./file')`, similar to
+how the `.cjs` extension is mandatory in `require` calls in CommonJS files.
+
+The `tsconfig.json` option `allowImportingTsExtensions` will allow the
+TypeScript compiler `tsc` to type-check files with `import` specifiers that
+include the `.ts` extension.
```

Comment:
I think https://github.com/nodejs/loaders/issues/214 must be resolved for this PR to land. I am deeply concerned about ecosystem fragmentation if we encourage people to publish packages that will be incompatible with:
- Code that doesn't run with type stripping enabled
- The official TypeScript compiler/type checker.

Reply:
Other than `.js` => `.ts` extension that requires the flag `allowImportingTsExtensions`, and I'm **strongly** against supporting `.js` for ts imports,
I guess the only thing that would make this incompatible is the lack of support for ts only features.
If we supported transformation of those ts features I guess there no incompatibility

Reply:
Specific TS features are irrelevant. My concerns are with packages published to npm that can then be used by another developer.

Reply:
if the package contains a `.ts` it can be consumed by node regardless of the flavor since type-stripping removes this problem of supporting different ts variants.
If package maintainer decide to ship ts only and break support for all js users, this is a decision of the package maintaner